src/win.h

```diff
@@ -51,4 +53,26 @@ typedef UINT32 QOS_FLOWID, *PQOS_FLOWID;
 #	endif
 #endif
 
+namespace win {
+static inline std::string utf16To8(const std::wstring &wstr) {
```

Comment:
Why do we need a Windows-specific conversion function? Shouldn't the new library be able to do this just as well?

Reply:
Then why the Windows-specific function? 👀

Reply:
I realized later that we needed the third-party (header-only) library in the client anyway, for `LegacyPlugin.cpp`.

As for `HostWindows.cpp`, I kept the Windows-specific function otherwise all plugins would've depended on the external library.

Reply:
If the library is only used in the implementation of that class, downstream code doesn't depend on it (it's a private dependency in that case)

Reply:
Only if the class is compiled into a separate object. In our case it's compiled as part of the plugin.

Reply:
> Only if the class is compiled into a separate object.

which is the case as every translation unit ends up as an individual object file, no?

Reply:
Yes, what I meant is that the object is not part of a library that the plugin links to.

In short: the third-party library has to exist in order for EVERY plugin (that relies on `HostWindows.cpp`) to be compiled. It's something I've always wanted to avoid, in fact the various classes don't depend on anything other than the standard library and the OS' API.

Reply:
I really don't see the issue here. We embed the source code of that library in Mumble anyway. And is people are using `HostWindows`, they depend on our source code anyway :shrug:

Reply:
Don't see a reason why that would be required/useful. Most plugins don't even work when built separately if not compiled with the exact same compiler as Mumble itself.
If we want separately buildable plugins, we should create a separate repo and put them there. :shrug:

Reply:
To be clear: I wouldn't be opposed to rely on a library if it turned out to be necessary, but in this case we don't gain any relevant advantages.

Also, keep in mind that the functions offered by Windows are designed to work with the argument(s) that are commonly used throughout WinAPI.

`utfcpp` deliberately avoids `std::wstring` because it doesn't use the same encoding on all platforms. This is due to `wchar_t` being 2 bytes on Windows (UTF-16) and 4 bytes on UNIX (UTF-32).

Basically, we don't end up with cleaner code as casting is required.

Reply:
> Don't see a reason why that would be required/useful. Most plugins don't even work when built separately if not compiled with the exact same compiler as Mumble itself. If we want separately buildable plugins, we should create a separate repo and put them there. 🤷

> Most plugins don't even work when built separately if not compiled with the exact same compiler as Mumble itself.

I wouldn't go as far as saying "exact same", but ABI compatibility is indeed required. It's one of the reasons we chose to write the new API in C rather than C++.

> If we want separately buildable plugins, we should create a separate repo and put them there. 🤷

That would be ideal.

Reply:
My point is that I would like to minimize the platform specific code in our code base as much as possible as that increases maintenance overhead.

Reply:
Only if said code is required on multiple platforms. In this case it's a Windows-only conversion that is necessary.